src/workerd/api/streams/readable-source.c++

```diff
@@ -0,0 +1,817 @@
+        // Note that we're passing amountToRead as the *minBytes* here so the tryRead should
+        // attempt to fill the entire buffer. If it doesn't, the implication is that we read
+        // everything.
+        size_t amount = co_await input.read(bytes.asBytes(), amountToRead);
+        KJ_DASSERT(amount <= amountToRead);
+
+        runningTotal += amount;
+        JSG_REQUIRE(runningTotal < limit, TypeError, "Memory limit exceeded before EOF.");
+
+        if (amount < amountToRead) {
+          // The stream has indicated that we're all done by returning a value less than the
+          // full buffer length.
+          // It is possible/likely that at least some amount of data was written to the buffer.
+          // In which case we want to add that subset to the parts list here before we exit
+          // the loop.
+          if (amount > 0) {
+            parts.add(bytes.first(amount).attach(kj::mv(bytes)));
+          }
+          break;
+        }
+
+        // Because we specify minSize equal to maxSize in the tryRead above, we should only
+        // get here if the buffer was completely filled by the read. If it wasn't completely
+        // filled, that is an indication that the stream is complete which is handled above.
+        KJ_DASSERT(amount == bytes.size());
+        parts.add(kj::mv(bytes));
+
+        // If the stream provided an expected length and our running total is equal to or
+        // greater than that length then we'll adjust our allocation strategy to be more
+        // conservative since the next read is likely (hopefully) going to be zero-length.
+        // Worst case, the stream is lying about its length and we have to keep reading.
+        // Best case, the next read is zero-length and we only had to do one additional
+        // small allocation.
+        KJ_IF_SOME(length, maybeLength) {
+          if (runningTotal >= length) {
+            amountToRead = kj::min(MIN_BUFFER_CHUNK, amountToRead);
+            continue;
+          }
+        }
+      }
+    }
+
+    if (option == ReadOption::NULL_TERMINATE) {
+      auto out = kj::heapArray<T>(runningTotal + 1);
+      out[runningTotal] = '\0';
+      copyInto<T>(out, parts);
+      co_return kj::mv(out);
+    }
+
+    // As an optimization, if there's only a single part in the list, we can avoid
+    // further copies.
+    if (parts.size() == 1) {
+      co_return kj::mv(parts[0]);
+    }
+
+    auto out = kj::heapArray<T>(runningTotal);
+    copyInto<T>(out, parts);
+    co_return kj::mv(out);
+  }
+
+  template <typename T>
+  void copyInto(kj::ArrayPtr<T> out, kj::ArrayPtr<kj::Array<T>> in) {
+    for (auto& part: in) {
+      KJ_DASSERT(part.size() <= out.size());
+      out.first(part.size()).copyFrom(part);
+      out = out.slice(part.size());
+    }
+  }
+};
+
+// An AsyncInputStream wrapper that translates tee-related kj::Exceptions from read
+// operations into jsg::Exceptions.
+class TeeErrorAdapter final: public kj::AsyncInputStream {
```

Comment:
Can we get rid of this if we define a KJ exception detail identifying tee errors appropriately? I recall `translateKjException()` pre-dates KJ exception details.